File: huipyo/src/week03/pgs42576/Solution.java
package week03.pgs42576;

import java.util.*;
import java.util.stream.Collectors;

class Solution {
    public String solution(String[] participant, String[] completion) {
        Map<String, List<String>> map = initMap(participant);

        for (String c : completion) {
            if (map.containsKey(c)) {
                map.get(c).remove(map.get(c).size() - 1);
            }
        }
        for (String key : map.keySet()) {
            if (!map.get(key).isEmpty()) return key;
        }
        return "";
    }

    private Map<String, List<String>> initMap(String[] participant) {
        return Arrays.stream(participant).collect(Collectors.groupingBy(s -> s));
    }
}

File: huipyo/src/week03/pgs42576/Solution.js
function solution(participant, completion) {
    let map = initMap(participant);

    for (let c of completion) {
        if (map.has(c)) {
            map.get(c).pop();
            if (map.get(c).length === 0) {
                map.delete(c);
            }
        }
    }
    return [...map.entries()].filter((k, v) => v.length != 0)[0][0];
}

function initMap(participant) {
    let map = new Map();

    participant.forEach(p => {
        if (map.has(p)) map.get(p).push(p);
        else map.set(p, [p]);
    });
    return map;
}